routing/router.go

```diff
@@ -1620,131 +1665,208 @@ func (r *ChannelRouter) sendPayment(payment *LightningPayment,
-			return [32]byte{}, nil, err
-		}
+			// Attempt to send this payment through the network to complete
+			// the payment. If this attempt fails, then we'll continue on
+			// to the next available route.
+			firstHop := lnwire.NewShortChanIDFromInt(
+				route.Hops[0].ChannelID,
+			)
 
-		err = r.cfg.Payer.SendHTLC(
-			firstHop, paymentID, htlcAdd,
-		)
-		if err != nil {
-			log.Errorf("Failed sending attempt %d for payment "+
-				"%x to switch: %v", paymentID, paymentHash, err)
+			// We generate a new, unique payment ID that we will use for
+			// this HTLC.
+			paymentID, err := r.cfg.NextPaymentID()
+			if err != nil {
+				return [32]byte{}, nil, err
+			}
 
-			// We must inspect the error to know whether it was
-			// critical or not, to decide whether we should
-			// continue trying.
-			finalOutcome := r.processSendError(
-				paySession, route, err,
-			)
-			if finalOutcome {
+			// We now have all the information needed to populate
+			// the current attempt information.
+			attempt = &channeldb.PaymentAttemptInfo{
+				PaymentID:  paymentID,
+				SessionKey: sessionKey,
+				Route:      *route,
+			}
+
+			// Before sending this HTLC to the switch, we checkpoint the
+			// fresh paymentID and route to the DB. This lets us know on
+			// startup the ID of the payment that we attempted to send,
+			// such that we can query the Switch for its whereabouts. The
+			// route is needed to handle the result when it eventually
+			// comes back.
+			err = r.cfg.Control.RegisterAttempt(payment.PaymentHash, attempt)
+			if err != nil {
 				return [32]byte{}, nil, err
 			}
 
-			lastError = err
-			continue
+			log.Tracef("Attempting to send payment %x (pid=%v), "+
+				"using route: %v", paymentHash, paymentID,
+				newLogClosure(func() string {
+					return spew.Sdump(route)
+				}),
+			)
+
+			// Send it to the Switch. When this method returns we assume
+			// the Switch successfully has persisted the payment attempt,
+			// such that we can resume waiting for the result after a
+			// restart.
+			err = r.cfg.Payer.SendHTLC(
+				firstHop, paymentID, htlcAdd,
+			)
+			if err != nil {
+				log.Errorf("Failed sending attempt %d for payment "+
+					"%x to switch: %v", paymentID, paymentHash, err)
+
+				// We must inspect the error to know whether it was
+				// critical or not, to decide whether we should
+				// continue trying.
+				finalOutcome := r.processSendError(
+					paySession, route, err,
+				)
+				if finalOutcome {
+					return [32]byte{}, nil, err
+				}
+
+				lastError = err
+
+				// We make another payment attempt.
+				attempt = nil
+				continue
+			}
+
+			log.Debugf("Payment %x (pid=%v) successfully sent to switch",
+				paymentHash, paymentID)
 		}
 
 		// Using the created circuit, initialize the error decrypter so we can
 		// parse+decode any failures incurred by this payment within the
 		// switch.
+		_, circuit, err := generateSphinxPacket(
+			&attempt.Route, paymentHash[:], attempt.SessionKey,
```

Comment:
We can avoid regenerating it by store it as a private field of the attempt struct or just caching the circuit and any related state in a loop-level variable above.

Reply:
Fixed: https://github.com/lightningnetwork/lnd/pull/2761/commits/48ee36937cfd1a6fb8ab2f16f715cf3f0779de55